File: 337-house-robber-iii/337-house-robber-iii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rob(self, root: Optional[TreeNode]) -> int:
        return max(self.dfs(root))
    def dfs(self, root):
        if root == None:
            return 0,0
        if root.left == None and root.right == None:
            # print(root.val)
            # print([root.val,0])
            return [root.val, 0]
        else:
            left_sum,max_c_ccl = self.dfs(root.left)
            right_sum, max_c_ccr = self.dfs(root.right)
            max_c_cc = max(left_sum+right_sum, left_sum+max_c_ccr, right_sum+max_c_ccl, max_c_ccr+max_c_ccl)
            return [root.val+max_c_ccl+max_c_ccr, max_c_cc]

File: Problems/301-400/337-house-robber-iii/337-house-robber-iii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int rob(TreeNode root) {
        int[] ans = this.dfs(root);
        return Math.max(ans[0], ans[1]);
    }
    public int[] dfs(TreeNode root){
        if (root==null) 
            return new int[]{0,0};
        if (root.left == null && root.right == null)
            return new int[]{root.val,0};
        else{
            int[] l = this.dfs(root.left);
            int[] r = this.dfs(root.right);
            int first = root.val + l[1] + r[1];
            int second = Math.max(l[0]+r[0], Math.max(l[1]+r[0], Math.max(l[0]+r[1], l[1]+r[1])));
            return new int[]{first, second};
        }
    }
 
            
}